libs/@local/harpc/server/src/router.rs

```diff
@@ -0,0 +1,244 @@
+use alloc::sync::Arc;
+use core::{
+    future::{self, Ready},
+    task::{Context, Poll},
+};
+
+use frunk::{HCons, HNil};
+use futures::future::Either;
+use harpc_net::codec::ErrorEncoder;
+use harpc_service::delegate::ServiceDelegate;
+use harpc_tower::{
+    body::{Body, BodyExt},
+    request::Request,
+    response::{BoxedResponse, Parts, Response},
+};
+use harpc_types::{service::ServiceId, version::Version};
+use tower::{Layer, Service, ServiceBuilder, ServiceExt, layer::util::Identity, util::Oneshot};
+
+use crate::{delegate::ServiceDelegateHandler, error::NotFound, session::SessionStorage};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handler<S> {
+    service: ServiceId,
+    version: Version,
+
+    inner: S,
+}
+
+pub trait Route<B, C> {
+    type ResponseBodyError;
+    type Future: Future<Output = Result<BoxedResponse<Self::ResponseBodyError>, !>> + Send;
+
+    fn call(&self, request: Request<B>, codec: C) -> Self::Future
+    where
+        B: Body<Control = !, Error: Send + Sync> + Send + Sync,
+        C: ErrorEncoder + Send + Sync;
+}
+
+// The clone requirement might seem odd here, but is the same as in axum's router implementation.
+// see: https://docs.rs/axum/latest/src/axum/routing/route.rs.html#45
+impl<B, C, E, Svc, Tail> Route<B, C> for HCons<Handler<Svc>, Tail>
+where
+    Svc: Service<Request<B>, Response = BoxedResponse<E>, Error = !, Future: Send>
+        + Clone
+        + Send
+        + Sync,
+    Tail: Route<B, C, ResponseBodyError = E> + Send,
+    B: Send,
+{
+    type Future = Either<Oneshot<Svc, Request<B>>, Tail::Future>;
+    type ResponseBodyError = E;
+
+    fn call(&self, request: Request<B>, codec: C) -> Self::Future
+    where
+        B: Body<Control = !, Error: Send + Sync> + Send + Sync,
+        C: ErrorEncoder + Send + Sync,
+    {
+        let requirement = self.head.version.into_requirement();
+
+        if self.head.service == request.service().id
+            && requirement.compatible(request.service().version)
+        {
+            let service = self.head.inner.clone();
+
+            Either::Left(service.oneshot(request))
+        } else {
+            Either::Right(self.tail.call(request, codec))
+        }
+    }
```

Comment:
Just so I understand correctly: Is this a list iteration and if we found the service we `oneshot` the request, otherwise we go to the next one and error if we hit `HNil`? Could you add a short comment please?

Reply:
see: [1e966ef](https://github.com/hashintel/hash/pull/5320/commits/1e966efcdacabb7210df7f00d2f6e94f25168973)